### PR TITLE
#76: Corrected and updated generate udf path method

### LINKED ISSUE #76: Add tests for the fixed generating bucket udf path method
This ticket https://github.com/exasol/bucketfs-utils-python/issues/74 does not include any test case for the bug it fixed.

### PR DESCRIPTION
**Pull request by umitbuyuksahin**
Fixes #76

### CODE REVIEW

File: exasol_bucketfs_utils_python/bucketfs_location.py [hunk 1]
```diff
@@ -35,6 +35,10 @@ def generate_bucket_udf_path(
     def get_complete_file_path_in_bucket(
             self,
             bucket_file_path: str) -> str:
+
+        if bucket_file_path is not None:
```
Comment (tkilias) [inline exasol_bucketfs_utils_python/bucketfs_location.py:39]: what do you do, if it is None, does PurePosixPath ignore it?

Comment (tkilias) [reply exasol_bucketfs_utils_python/bucketfs_location.py:39]: you need to add a test parameter for None

File: tests/integration_tests/without_db/test_bucketfs_location.py [hunk 2]
```diff
@@ -24,13 +24,22 @@ def test_generate_bucket_udf_path(path_in_bucket):
...
         bucket_name="default", bucketfs_config=bucketfs_config)
-    bucketfs_location = BucketFSLocation(bucket_config, "")
 
-    udf_path = bucketfs_location.generate_bucket_udf_path(
-        path_in_bucket=path_in_bucket)
+    base_path = "base_path"
+    bucketfs_location_with_base_path = \
+        BucketFSLocation(bucket_config, base_path=PurePosixPath(base_path))
+    bucketfs_location_without_base_path = \
+        BucketFSLocation(bucket_config, base_path=PurePosixPath(""))
```
Comment (tkilias) [inline tests/integration_tests/without_db/test_bucketfs_location.py:32]: can be the base_path None?

Comment (tkilias) [inline tests/integration_tests/without_db/test_bucketfs_location.py:32]: I suggest to split with and without base_path into two tests, makes it more obvious

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol_bucketfs_utils_python/bucketfs_location.py (hunk 1 above)
Comment (umitbuyuksahin) [reply exasol_bucketfs_utils_python/bucketfs_location.py:39]: PurePosixPath raises error for null input

File: tests/integration_tests/without_db/test_bucketfs_location.py (hunk 2 above)
Comment (umitbuyuksahin) [reply tests/integration_tests/without_db/test_bucketfs_location.py:32]: base_path shouldn't be None.
it might be better to make it empty string in case of None input, should we ?

Comment (tkilias) [reply tests/integration_tests/without_db/test_bucketfs_location.py:32]: Actually, it should be an optional parameter and as this None, at least that what I intended.

File: exasol_bucketfs_utils_python/bucketfs_location.py (hunk 1 above)
Comment (tkilias) [reply exasol_bucketfs_utils_python/bucketfs_location.py:39]: than you need to handle None

File: tests/integration_tests/without_db/test_bucketfs_location.py [hunk 3]
```diff
@@ -17,20 +92,21 @@
     "path/in/bucket/file.txt.zip",
     "path/in/bucket/file.txt.tgz",
     "path/in/bucket/file.txt.tar"])
-def test_generate_bucket_udf_path(path_in_bucket):
+def test_generate_bucket_udf_path_with_base_path(path_in_bucket):
```
Comment (tkilias) [inline tests/integration_tests/without_db/test_bucketfs_location.py:95]:
```suggestion
def test_generate_bucket_udf_path_without_base_path(path_in_bucket):
```

File: exasol_bucketfs_utils_python/bucketfs_location.py [hunk 4]
```diff
@@ -21,8 +21,9 @@ class BucketFSLocation(AbstractBucketFSLocation):
     object type is known inside the UDF.
     """
 
-    def __init__(self, bucket_config: BucketConfig, base_path: PurePosixPath):
-        self.base_path = base_path
+    def __init__(self, bucket_config: BucketConfig,
```
Comment (tkilias) [inline exasol_bucketfs_utils_python/bucketfs_location.py:24]: Did you change the base_path handling also in the MockBucketFSLocation?

File: doc/changes/changes_0.4.0.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# BucketFs Utils Python 0.4.0, released YYYY-MM-DD
+Code name:  TBD.
+
+## Summary
+TBD.
+
+## Features / Enhancements
+
+ - #72: Added tests for the fixed generating bucket udf path method
```
Comment (tkilias) [inline doc/changes/changes_0.4.0.md:9]: Mention the change of the handling of None for the base_path

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol_bucketfs_utils_python/localfs_mock_bucketfs_location.py [hunk 6]
```diff
@@ -16,10 +16,15 @@ class LocalFSMockBucketFSLocation(AbstractBucketFSLocation):
     or to download or read objects into strings, fileobjects or joblib objects from the LocalFS given a file path.
     """
 
-    def __init__(self, base_path: PurePosixPath):
-        self.base_path = base_path
+    def __init__(self, base_path: Optional[PurePosixPath]):
+        self.base_path = "" if base_path is None else base_path
 
     def get_complete_file_path_in_bucket(self, bucket_file_path) -> str:
```
Comment (tkilias) [inline exasol_bucketfs_utils_python/localfs_mock_bucketfs_location.py:22]: Type for bucket_file_path is missing

File: exasol_bucketfs_utils_python/bucketfs_location.py [hunk 7]
```diff
@@ -35,6 +36,12 @@ def generate_bucket_udf_path(
     def get_complete_file_path_in_bucket(
             self,
             bucket_file_path: str) -> str:
```
Comment (tkilias) [inline exasol_bucketfs_utils_python/bucketfs_location.py:38]: wrong type, we should use PurePosixPath

File: exasol_bucketfs_utils_python/localfs_mock_bucketfs_location.py (hunk 6 above)
Comment (tkilias) [reply exasol_bucketfs_utils_python/localfs_mock_bucketfs_location.py:22]: we also should use PurePosixPath

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol_bucketfs_utils_python/bucketfs_location.py (hunk 7 above)
Comment (umitbuyuksahin) [reply exasol_bucketfs_utils_python/bucketfs_location.py:38]: it is used as string in almost entire the project

**Review by tkilias — APPROVED** [review]
